### PR TITLE
feat: full memtrie logic for range retain

### PR DESCRIPTION
**Pull request by Longarithm**
Next step on #12074.
Supporting all cases I came up with where nodes restructuring is required.
Surprisingly, `squash_node` is enough to call. I only needed to implement trivial cases which weren't possible for single key deletion.
I implemented the tests first, and majority of them failed before changing the logic.
Each test is comparing naive approach with `retain_multi_range`.

### Notes
* I'm a bit scared that I didn't realise the need to squash Extension before. Fortunately, `squash_node` handles that, but if you feel some cases are not covered here, feel free to post suggestions!
* Reused + copypasted some Robin' tooling to generate interesting nodes conversions.
* Note that test for reading "extra" child node is not required because we always read all children.

### Next steps
* A bit more testing
* Similar logic for partial trie
* Generating intervals needed for resharding
* Use that to implement shard switch on chain

### CODE REVIEW

**Review by Trisfald — APPROVED** [review]
LGTM

File: core/store/src/trie/mem/nibbles_utils.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+pub(crate) fn hex_to_nibbles(hex: &str) -> Vec<u8> {
+    if hex == "_" {
+        return vec![];
+    }
+    assert!(hex.len() % 2 == 0);
+    hex::decode(hex).unwrap()
```
Comment (wacban) [inline core/store/src/trie/mem/nibbles_utils.rs:6]: Is that right? Hex decodes to bytes but nibbles are half-bytes. I do realize you only moved it and it's probably ok so I'm mostly curious where is the gap in my understanding.

File: core/store/src/trie/mem/nibbles_utils.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+pub(crate) fn all_nibbles(hexes: &str) -> Vec<Vec<u8>> {
```
Comment (wacban) [inline core/store/src/trie/mem/nibbles_utils.rs:9]: nit: how about something like `multi_hex_to_nibbles`?

File: core/store/src/trie/mem/nibbles_utils.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    hexes.split_whitespace().map(|x| hex_to_nibbles(x)).collect()
+}
+
+pub(crate) fn full_16ary_tree_keys() -> Vec<Vec<u8>> {
```
Comment (wacban) [inline core/store/src/trie/mem/nibbles_utils.rs:13]: nit: How about `all_two_nibble_nibbles` or `all_eight_bit_nibbles`?

File: core/store/src/trie/mem/resharding.rs [hunk 4]
```diff
@@ -140,19 +138,17 @@ impl<'a, M: ArenaMemory> MemTrieUpdate<'a, M> {
...
-                        },
-                    );
-                }
+                self.place_node(
+                    node_id,
+                    UpdatedMemTrieNode::Extension {
+                        extension,
+                        child: OldOrUpdatedNodeId::Updated(new_child_id),
+                    },
+                );
```
Comment (wacban) [inline core/store/src/trie/mem/resharding.rs:147]: nit: I think it would be more readable multilined

File: core/store/src/trie/mem/updating.rs [hunk 5]
```diff
@@ -518,66 +521,65 @@ impl<'a, M: ArenaMemory> MemTrieUpdate<'a, M> {
...
-                UpdatedMemTrieNode::Branch { mut children, value } => {
-                    // Remove any children that are now empty (removed).
-                    for child in children.iter_mut() {
-                        if let Some(OldOrUpdatedNodeId::Updated(child_node_id)) = child {
-                            if let UpdatedMemTrieNode::Empty =
-                                self.updated_nodes[*child_node_id as usize].as_ref().unwrap()
-                            {
-                                *child = None;
-                            }
+    pub(crate) fn squash_node(&mut self, node_id: UpdatedMemTrieNodeId) {
```
Comment (wacban) [inline core/store/src/trie/mem/updating.rs:524]: Can you update the comment please?

File: core/store/src/trie/mem/resharding.rs [hunk 6]
```diff
@@ -140,19 +138,17 @@ impl<'a, M: ArenaMemory> MemTrieUpdate<'a, M> {
...
             }
         }
+
+        self.squash_node(node_id);
```
Comment (wacban) [inline core/store/src/trie/mem/resharding.rs:151]: The correctness of squashing depends on it being called on nodes in the correct order. 

sanity check: Is it called in the right order here? 

nit: Can you add a comment about it for future us?

File: core/store/src/trie/mem/resharding.rs [hunk 7]
```diff
@@ -222,27 +276,234 @@ mod tests {
...
-        let memtrie_changes = update.to_mem_trie_changes_only();
-        let state_root = memtries.apply_memtrie_changes(0, &memtrie_changes);
+        key
+    }
 
-        let update = memtries.update(state_root, true).unwrap();
-        let (mut trie_changes, _) = update.retain_multi_range(&[retain_range]);
-        let memtrie_changes = trie_changes.mem_trie_changes.take().unwrap();
-        let new_state_root = memtries.apply_memtrie_changes(1, &memtrie_changes);
+    fn check_random(max_key_len: usize, max_keys_count: usize, test_count: usize) {
```
Comment (wacban) [inline core/store/src/trie/mem/resharding.rs:459]: nice

**Review by wacban — APPROVED** [review]
Looks great! 

It's very elegant, I like it a lot that we can just squash at the end. 

I didn't check all of the tests but the following transitions seem possible and test worthy:
empty -> empty
leaf -> empty / leaf
extension -> empty / leaf / extension / branch
branch -> empty / leaf / extension / branch

File: core/store/src/trie/mem/nibbles_utils.rs (hunk 1 above)
Comment (Longarithm) [reply core/store/src/trie/mem/nibbles_utils.rs:6]: I think the Robin's assumption was that we can only write actual bytes and we can't write key with odd number of nibbles. Maybe trie internals shouldn't be aware of that; on the other hand, vectors of even lengths cover all cases well.

Comment (wacban) [reply core/store/src/trie/mem/nibbles_utils.rs:6]: I was expecting the bytes to be further split down into nibbles, so something like this:
```
hex::decode(hex).unwrap().iter().map(|byte| vec![byte & 0xf0 >> 4, byte & 0x0f]).collect_vec().concat()
```

Comment (Longarithm) [reply core/store/src/trie/mem/nibbles_utils.rs:6]: No, each string symbol represents a readable nibble (0..9, a..f). I'll add comment

**Comment by Longarithm** [comment]
Sanity check:

empty -> empty

checked everywhere if at least 1 node is removed

leaf -> empty / leaf

one of the simplest, test_retain_single_range. depends if leaf is retained or not

extension -> empty

test_descend_and_remove

extension -> leaf

test_extend_extensions

extension -> extension

actually was tricky.
squashing happens in test_prefixes
there is a sequence ext -> branch -> ext, branch is removed, two exts are merged.

branch -> empty

test_descend_and_remove

branch -> leaf

test_branch_to_leaf

branch -> extension

test_branch_to_extension

branch -> branch

many of them, e.g. test_full_16ary

**Pull request merged**